src/ddmd/hdrgen.d

```diff
@@ -1855,7 +1855,18 @@ public:
         // Don't print tf.mod, tf.trust, and tf.linkage
         if (!f.inferRetType && tf.next)
             typeToBuffer(tf.next, null);
+
         parametersToBuffer(tf.parameters, tf.varargs);
+
+        if ((hgs.hdrgen || hgs.fullDump) && (!tf.next || f.inferRetType))
```

Comment:
Looks like you try to distinguish inferred attributes from given ones, but doing that depending on the return type seems fragile. Any other method to preserve the given attributes?